File: core/documents/heading_engine.py
from __future__ import annotations

import re
from typing import Dict, Any


def extract_headings(text: str) -> Dict[str, Any]:
    src = text or ""
    md = [{"level": len(h), "title": t.strip()} for h, t in re.findall(r"^(#{1,6})\s+(.+)$", src, flags=re.MULTILINE)]
    html = [{"level": int(n), "title": t.strip()} for n, t in re.findall(r"<h([1-6])[^>]*>(.*?)</h\1>", src, flags=re.IGNORECASE|re.DOTALL)]
    headings = sorted(md + html, key=lambda x: (x["level"], x["title"]))
    return {"headings": headings}

File: src/documents/headingEngine.ts
/**
 * Converted from Python: core/documents/heading_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function extractHeadings(text: any): any {
  var src: any = py.or2(text, () => (""));
  var md: any = py.iter(py.reFindall("^(#{1,6})\\s+(.+)$", src, "m")).map(([h, t]: any) => ({"level": py.len(h), "title": py.strip(t)}));
  var html: any = py.iter(py.reFindall("<h([1-6])[^>]*>(.*?)</h\\1>", src, "is")).map(([n, t]: any) => ({"level": py.toInt(n), "title": py.strip(t)}));
  var headings: any = py.sorted(py.add(md, html), {key: ((x: any) => [py.at(x, "level"), py.at(x, "title")]) as (item: any) => any});
  return {"headings": headings};
}
